File: compiler/nodes/patterns/wildcard.go
package patterns

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/visit"
)

type WildcardPatternNode struct {
	Facts *database.Facts

	matching database.Node
}

func (node *WildcardPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseWildcardPattern(parser *syntax.Parser) (*WildcardPatternNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("UnderscoreKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this wildcard pattern")

	return &WildcardPatternNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *WildcardPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node
}

func (node *WildcardPatternNode) Codegen(c *codegen.Codegen) error {
	// No code needed
	return nil
}

File: compiler/src/nodes/patterns/wildcard.rs
use crate::{
    codegen::{Codegen, CodegenCtx, ir},
    database::{Node, NodeRef},
    nodes::visit_pattern,
    syntax::{ParseError, Parser, TokenKind},
    visit::{MatchPathSegment, Visit, Visitor},
};

#[derive(Debug)]
pub struct WildcardPatternNode;

impl Node for WildcardPatternNode {}

pub fn parse_wildcard_pattern(parser: &mut Parser<'_>) -> Result<WildcardPatternNode, ParseError> {
    parser.token(TokenKind::UnderscoreKeyword)?;
    parser.commit("in this wildcard pattern");

    Ok(WildcardPatternNode)
}

impl Visit for WildcardPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, Some(MatchPathSegment::Match));
    }
}

impl Codegen for WildcardPatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> Option<ir::SpannedExpression> {
        ir::Expression::And(Vec::new()).at(node, ctx)
    }
}
